jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1023,95 +1125,80 @@ public void run()
-
                     try
                     {
-                        // Look for an immediately available job
-                        job = _jobs.poll();
-                        if (job == null)
+                        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+                        Runnable job = idleJobPoll(idleTimeoutNanos);
+
+                        while (job != null)
                         {
-                            // No job immediately available maybe we should shrink?
-                            long idleTimeout = getIdleTimeout();
-                            if (idleTimeout > 0 && getThreads() > _minThreads)
-                            {
-                                long last = _lastShrink.get();
-                                long now = NanoTime.now();
-                                if (NanoTime.millisElapsed(last, now) > idleTimeout && _lastShrink.compareAndSet(last, now))
-                                {
-                                    if (LOG.isDebugEnabled())
-                                        LOG.debug("shrinking {}", QueuedThreadPool.this);
-                                    break;
-                                }
-                            }
-
-                            // Wait for a job, only after we have checked if we should shrink
-                            job = idleJobPoll(idleTimeout);
-
-                            // If still no job?
-                            if (job == null)
-                                // continue to try again
-                                continue;
+                            idle = false;
+                            // Run the jobs.
+                            if (LOG.isDebugEnabled())
+                                LOG.debug("run {} in {}", job, QueuedThreadPool.this);
+                            doRunJob(job);
+                            if (LOG.isDebugEnabled())
+                                LOG.debug("ran {} in {}", job, QueuedThreadPool.this);
+
+                            // Signal that we are idle again; since execute() subtracts
+                            // 1 from idle each time a job is submitted, we have to add
+                            // 1 for each executed job here to compensate.
+                            if (!addCounts(0, 1))
+                                break;
+                            idle = true;
+
+                            // Look for another job
+                            job = _jobs.poll();
                         }
 
-                        idle = false;
-
-                        // run job
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("run {} in {}", job, QueuedThreadPool.this);
-                        runJob(job);
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("ran {} in {}", job, QueuedThreadPool.this);
+                        shrunk = shrink();
```

Comment:
If thread reaches here after doing `one job immediately`, then it might shirk that thread. And that may cause further thread-creation and thread-exit pattern (thrashing)

I think bigger question here is do we want to make sure thread remain idle for `idletimeout` (or keep-alive time) ?

Reply:
I think a minimum time to live off one idle timeout will stop and thrashing.

Reply:
Actually, I have just re-added the behaviour that when we start a new thread we advance the shrink threshold to now, so no threads will be able to shrink immediately after one has started.